cuda/solver/gmres_kernels.cu

```diff
@@ -89,12 +436,126 @@ void step_1(std::shared_ptr<const CudaExecutor> exec,
             const matrix::Dense<ValueType> *b_norm, const size_type iter,
             const Array<stopping_status> *stop_status)
 {
-    NOT_IMPLEMENTED;
+    finish_arnoldi(exec, next_krylov_basis, krylov_bases, hessenberg_iter, iter,
+                   stop_status->get_const_data());
+    givens_rotation(exec, givens_sin, givens_cos, hessenberg_iter,
+                    residual_norm, residual_norms, b_norm, iter, stop_status);
 }
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_GMRES_STEP_1_KERNEL);
 
 
+namespace kernel {
+
+
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void solve_upper_triangular(
+    size_type num_cols, size_type num_rhs,
+    const ValueType *__restrict__ residual_norms,
+    ValueType *__restrict__ hessenberg, ValueType *__restrict__ y,
+    const size_type *__restrict__ final_iter_nums)
+{
+    const auto local_id = thread::get_local_thread_id<cuda_config::warp_size>();
+
+    if (local_id >= num_rhs) return;
+
+    for (int i = final_iter_nums[local_id] - 1; i >= 0; --i) {
+        auto temp = residual_norms[i * num_rhs + local_id];
+        for (size_type j = i + 1; j < final_iter_nums[local_id]; ++j) {
+            temp -= hessenberg[i * num_cols + j * num_rhs + local_id] *
+                    y[j * num_rhs + local_id];
+        }
+        y[i * num_rhs + local_id] =
+            temp / hessenberg[i * num_cols + i * num_rhs + local_id];
+        __syncthreads();
+    }
+    // Solve upper triangular.
+    // y = hessenberg \ residual_norms
+}
+
+
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void calculate_Qy(
+    const size_type num_rows, const size_type num_cols, const size_type num_rhs,
+    const ValueType *__restrict__ krylov_bases, const ValueType *__restrict__ y,
+    ValueType *__restrict__ before_preconditioner,
+    const size_type *__restrict__ final_iter_nums)
```

Comment:
Add a stride parameter and take it into consideration when accessing the values.